### PR TITLE
Add support for arbitrary price calculators based on price definitions.

### PR DESCRIPTION
**Pull request by kleinmann**
Unfortunately the `QuantityPriceCalculator` currently takes on a rather special role in the whole application, so I thought it more prudent to enable arbitrary calculators first and leave a refactoring for "later" ;-)

### CODE REVIEW

File: src/Core/Checkout/Cart/Calculator.php [hunk 1]
```diff
@@ -25,23 +22,16 @@ class Calculator
...
-     * @var AbsolutePriceCalculator
-     */
-    private $absolutePriceCalculator;
+    private $calculators;
 
     public function __construct(
         QuantityPriceCalculator $priceCalculator,
-        PercentagePriceCalculator $percentagePriceCalculator,
-        AbsolutePriceCalculator $absolutePriceCalculator
+        iterable $calculators
```
Comment (bcremer) [inline src/Core/Checkout/Cart/Calculator.php:35]: `$calculators` must be annotated in phpdoc so the Static Code Analysis can reason about the constructor:

```
    /**
     * @param PriceCalculatorInterface[] $calculators
     */
```

File: src/Core/Checkout/Cart/Price/PercentagePriceCalculator.php [hunk 2]
```diff
@@ -35,19 +37,26 @@ public function __construct(
...
+        return $priceDefinition instanceof PercentagePriceDefinition;
+    }
+
     /**
      * Provide a negative percentage value for discount or a positive percentage value for a surcharge
      *
-     * @param float           $percentage 10.00 for 10%, -10.0 for -10%
-     * @param PriceCollection $prices
-     * @param CheckoutContext $context
+     * @param PercentagePriceDefinition $priceDefinition
```
Comment (bcremer) [inline src/Core/Checkout/Cart/Price/PercentagePriceCalculator.php:48]: The `@param` are superfluous. The native type definitions provide all required type information.

File: src/Core/Checkout/Cart/Price/Struct/PriceDefinitionInterface.php [hunk 3]
```diff
@@ -2,6 +2,9 @@
 
 namespace Shopware\Core\Checkout\Cart\Price\Struct;
 
+use Shopware\Core\Framework\Rule\Rule;
+
 interface PriceDefinitionInterface
 {
+    public function getFilter(): ?Rule;
```
Comment (bcremer) [inline src/Core/Checkout/Cart/Price/Struct/PriceDefinitionInterface.php:9]: Must this definition be nullable?

Maybe a `NullRull` should be returned instead.

File: src/Core/Checkout/Cart/Calculator.php [hunk 4]
```diff
@@ -103,20 +93,14 @@ private function calculatePrice(Cart $cart, LineItem $lineItem, CheckoutContext
 
         $definition = $lineItem->getPriceDefinition();
 
-        if ($definition instanceof AbsolutePriceDefinition) {
-            //reduce line items for provided filter
-            $prices = $this->filterLineItems($calculated, $definition->getFilter(), $context)
-                ->getPrices();
+        foreach ($this->calculators as $calculator) {
+            if ($calculator->supports($definition)) {
```
Comment (bcremer) [inline src/Core/Checkout/Cart/Calculator.php:101]: The first calculator that  supports the given `$definition` "wins".

Is this expected or should there only be exact one supported calculator per `calculate()` run.

Maybe throw an LogikException if more than one Calculator supports given `$definition`.

File: src/Core/Checkout/Cart/Price/AbsolutePriceCalculator.php [hunk 5]
```diff
@@ -26,11 +28,23 @@ public function __construct(QuantityPriceCalculator $priceCalculator, Percentage
...
+    }
+
+    /**
+     * @param AbsolutePriceDefinition $priceDefinition
+     * @param PriceCollection         $prices
+     * @param CheckoutContext         $context
+     *
+     * @return Price
+     */
+    public function calculate(PriceDefinitionInterface $priceDefinition, PriceCollection $prices, CheckoutContext $context): Price
```
Comment (bcremer) [inline src/Core/Checkout/Cart/Price/AbsolutePriceCalculator.php:43]: See: https://github.com/shopware/platform/pull/6/files?utf8=%E2%9C%93&diff=split#r235986018

File: src/Core/Checkout/Cart/Price/Struct/QuantityPriceDefinition.php [hunk 6]
```diff
@@ -63,4 +64,9 @@ public function setQuantity(int $quantity): void
     {
         $this->quantity = $quantity;
     }
+
+    public function getFilter(): ?Rule
+    {
+        return null;
```
Comment (bcremer) [inline src/Core/Checkout/Cart/Price/Struct/QuantityPriceDefinition.php:70]: See: https://github.com/shopware/platform/pull/6/files?utf8=%E2%9C%93&diff=split#r235986581

File: src/Core/Checkout/Cart/Price/Struct/PriceDefinitionInterface.php (hunk 3 above)
Comment (kleinmann) [reply src/Core/Checkout/Cart/Price/Struct/PriceDefinitionInterface.php:9]: Currently there is no concept of a `NullRule` in the rule system, as far as I can tell. In this particular case of the price calculation the entire rule verification is circumvented if no rule is set. This matches my assumptions, as a no-op evaluation cycle of the rule system seems rather odd to me.
I'd love to see someone from the core development team chime in with their thoughts on this :-)

File: src/Core/Checkout/Cart/Calculator.php (hunk 4 above)
Comment (kleinmann) [reply src/Core/Checkout/Cart/Calculator.php:101]: Multiple calculators per pricing definition is a valid use-case for me. Think along the lines of "custom calculator for core definitions via plugin" (although a custom pricing definition may be more suited). A developer can then modify the actually used calculator by using `priority` in their service's tag definition to change the order in the tagged service collection.

File: src/Core/Checkout/Cart/Calculator.php (hunk 1 above)
Comment (kleinmann) [reply src/Core/Checkout/Cart/Calculator.php:35]: Good point, added the docblock.

File: src/Core/Checkout/Cart/Price/PercentagePriceCalculator.php (hunk 2 above)
Comment (kleinmann) [reply src/Core/Checkout/Cart/Price/PercentagePriceCalculator.php:48]: Actually the interface only defines a `PricingDefinitionInterface`. As we're using `$pricingDefinition->getPercentage()` here, the type definitions are too abstract. This also results in a static code analysis error:
```
Call to an undefined method Shopware\Core\Checkout\Cart\Price\Struct\PriceDefinitionInterface::getPercentage().
```

I prefer adding a full docblock even if I need to describe just one of the parameters.

**Comment by jenskueper** [comment]
Hello @kleinmann!
Thank you for the pull request! You are right that the current implementation is not ideal and hard to extend. I like the concept of having a registry with calculators which can define which PriceDefinition they support.

We are currently working on different features regarding the cart and orders. I would therefore suggest to wait until these changes are live and then we come back to your pull request and fix the calculator issue.

I hope this is okay for you!

@bcremer thank you for the comments and suggestions!

Best regards,
Jens

**Comment by kleinmann** [comment]
@jenskueper This is fine by me. Just give me a quick ping once we can continue here :-)

**Comment by mitelg** [comment]
hey @kleinmann 
thanks for your contribution. could you please sign the CLA?

**Comment by kleinmann** [comment]
@mitelg Sure, signed :-)

**Comment by mitelg** [comment]
btw, is this PR still needed? :smile:

**Comment by kleinmann** [comment]
@mitelg As far as I can tell `\Shopware\Core\Checkout\Cart\Calculator` is still kinda the same, only supporting the three core price definitions and not having an interface for extension 🙂
If there's not an alternative way of doing price calculation I'm currently missing, I'd argue it may still be a viable change.

It may also be possible by abusing the `AbsolutePriceDefinition` with some interface to add customized calculations in front of that. That's just a quick thought, though 🙂

**Comment by pantrtxp** [comment]
Given the age of this PR and that it was imported into an issue I'm going to close this PR

**Comment by mitelg** [comment]
@kleinmann is this even still relevant?

 if so, a new PR would probably be merged faster, than this old one :wink:

**Comment by kleinmann** [comment]
As my previous comment already said, yes, I think this is still relevant and needed. I'll look into creating a new PR.

**Comment by mitelg** [comment]
as the ticket is not closed, this is still in our scheduled tasks. don't get that wrong :blush: 
a new PR would be nice. maybe you could do it on the master (6.2) branch, as it would probably bring some breaks with it

**Pull request closed without merging**